lib/booking_system.rb

```diff
@@ -0,0 +1,118 @@
+      @reservations << reservation
+      # binding.pry
+    end
+
+    def make_block(start_date, end_date, number_of_rooms)
+      if number_of_rooms > 5
+        raise ArgumentError, "Cannot reserve more than 5 rooms"
+      end
+
+      id = generate_id
+
+      block = Hotel::Block.new(block_id: id, number_of_rooms: number_of_rooms, start_date: start_date, end_date: end_date, price_per_night: 150)
+      @blocks << block
+
+      number_of_rooms.times do
+        block_reservation = Hotel::BlockReservation.new(block_id: id,  reservation_id: nil, room: assign_available_room(start_date, end_date), start_date: start_date, end_date: end_date, price_per_night: 150)
+        @reservations << block_reservation
+      end
+    end
+
+    def make_block_reservation(block_id)
+      block_reservation = find_empty_block_reservation(block_id)
+      block_reservation.reservation_id = generate_id
+      return block_reservation
+    end
+
+    def assign_available_room(start_date, end_date)
+      booked_rooms = []
+      available_rooms = []
+      @reservations.each do |reservation|
+        if
+          reservation.start_date < end_date && start_date < reservation.end_date
+          booked_rooms << reservation.room
+        end
+      end
+      all_rooms = load_rooms
+      available_rooms = all_rooms - booked_rooms
+      if available_rooms[0] == nil
+        raise ArgumentError, "No rooms available"
+      end
+      return available_rooms[0]
+    end
+
+    def search_reservations(start_date, end_date)
+      reservations_within_date = []
+      @reservations.each do |reservation|
+        if
+          reservation.start_date < end_date && start_date < reservation.end_date
+          reservations_within_date << reservation
+        end
+      end
+      # binding.pry
+      return reservations_within_date
+    end
+
+    def find_reservation(reservation_id)
+      reservation = @reservations.find { |reservation| reservation.reservation_id == reservation_id }
+      return reservation
+    end
+
+    def find_empty_block_reservation(block_id)
+      block_reservation = @reservations.find { |block_reservation| block_reservation.block_id == block_id && block_reservation.reservation_id == nil}
+      if block_reservation == nil
+        raise ArgumentError, "all rooms in block have been reserved"
+      end
+      return block_reservation
+    end
+
+    # def find_block(block_id)
+    #   block = @reservations.find_all { |block_reservation| block_reservation.block_id == block_id }
+    #   return block
+    # end
+
+    def generate_id
+      reservation_id = rand(1..100000)
+      check_id(reservation_id)
+      @all_ids << reservation_id
+      return reservation_id
+    end
+
+    def check_id(reservation_id)
+      if @all_ids.include?(reservation_id)
+        raise ArgumentError, "reservation_id already exists"
+      end
```

Comment:
I don't know if raising an argument error is the right choice here, since the user didn't supply this ID, the program generated it. Instead you might regenerate the ID, or assign IDs sequentially so that there's no possibility of an collision (this is what Rails does).